# بسم الله الرحمن الرحيم

In [1]:
# import pandas as pd 
# import numpy as np 
# from matplotlib import pyplot as plt 
# import seaborn as sns 
# plt.style.use('default')
# #-------------------------------------------------------------------------------------
# from sklearn.utils import resample 
# from sklearn import preprocessing
# from sklearn.linear_model import LogisticRegression
# from sklearn.model_selection import train_test_split
# from sklearn.ensemble import RandomForestClassifier
# from sklearn.tree import DecisionTreeClassifier
# from sklearn.svm import SVC
# from sklearn.pipeline import make_pipeline
# from sklearn.metrics import confusion_matrix
# from sklearn.model_selection import cross_validate

# from sklearn.naive_bayes import GaussianNB
# from sklearn.metrics import confusion_matrix
# from sklearn.metrics import classification_report
# from sklearn.metrics import precision_score
# from sklearn.metrics import recall_score
# from sklearn.pipeline import Pipeline 

# import tensorflow_addons as tfa
# import tensorflow as tf 
# from tensorflow import keras 
# from tensorflow.keras import layers 
# from focal_loss import SparseCategoricalFocalLoss

# #-------------------------------------------------------------------------------------

# from keras.callbacks import EarlyStopping
# from keras.callbacks import ModelCheckpoint
# #-------------------------------------------------------------------------------------
# from imblearn.over_sampling import SMOTE
# from utils import *
# %matplotlib inline

In [2]:
from utils import *
import pandas as pd 
from sklearn.model_selection import train_test_split
import tensorflow as tf 
import tensorflow.keras 
import pickle 
import seaborn as sns 
from matplotlib import pyplot as plt 


2023-05-17 08:02:28.039159: I tensorflow/core/platform/cpu_feature_guard.cc:193] This TensorFlow binary is optimized with oneAPI Deep Neural Network Library (oneDNN) to use the following CPU instructions in performance-critical operations:  AVX2 FMA
To enable them in other operations, rebuild TensorFlow with the appropriate compiler flags.
2023-05-17 08:02:28.210008: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libcudart.so.11.0'; dlerror: libcudart.so.11.0: cannot open shared object file: No such file or directory
2023-05-17 08:02:28.210041: I tensorflow/compiler/xla/stream_executor/cuda/cudart_stub.cc:29] Ignore above cudart dlerror if you do not have a GPU set up on your machine.
2023-05-17 08:02:29.074899: W tensorflow/compiler/xla/stream_executor/platform/default/dso_loader.cc:64] Could not load dynamic library 'libnvinfer.so.7'; dlerror: libnvinfer.so.7: cannot open shared object file: No such file or directory
2023-

## Reading the Dataset, cleaning , preparing it for the training 

In [3]:
df = pd.read_csv("dataset/body_level_classification_train.csv")

In [4]:
df_original= cleaning_data(df).copy()

In [5]:
all_class_train,all_class_test=data_spliting(df_original,1/3)

In [6]:
X_train=all_class_train.drop(["Body_Level"],axis=1).to_numpy()
Y_train=all_class_train["Body_Level"].to_numpy()
X_test =all_class_test.drop(["Body_Level"],axis=1).to_numpy()
Y_test =all_class_test["Body_Level"].to_numpy()

In [7]:
X_train, scaler= data_scaling(X_train)
with open(f"scaler","wb") as fp : 
    pickle.dump(scaler,fp)

In [8]:
X_val, X_test, Y_val, Y_test = train_test_split(X_test
                                                     , Y_test, test_size=0.5, random_state=42)

In [9]:
X_val = scaler.transform(X_val)
X_test=scaler.transform(X_test)

In [11]:
model=build_model(2,0.01)

In [12]:
history = model.fit(X_train,Y_train,validation_data=[X_val,Y_val] 
             ,batch_size=32, epochs=300 , verbose =1, 
                       callbacks=[CustomCallback("NNModels/before_sampling/before_sampling.h5")])

Epoch 1/300
31/31 [==============================] - 1s 8ms/step - loss: 0.8218 - accuracy: 0.4817 - val_loss: 0.4839 - val_accuracy: 0.6204
Epoch 2/300
31/31 [==============================] - 0s 3ms/step - loss: 0.3955 - accuracy: 0.6907 - val_loss: 0.3360 - val_accuracy: 0.7061
Epoch 3/300
31/31 [==============================] - 0s 3ms/step - loss: 0.3060 - accuracy: 0.7525 - val_loss: 0.2871 - val_accuracy: 0.7714
Epoch 4/300
28/31 [==========================>...] - ETA: 0s - loss: 0.2541 - accuracy: 0.7924
Yes You are here 0.7870182394981384 0.7836734652519226
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                         

config.json                                    2023-05-17 08:02:47         1588
metadata.json                                  2023-05-17 08:02:47           64
31/31 [==============================] - 0s 4ms/step - loss: 0.1573 - accuracy: 0.8742 - val_loss: 0.1725 - val_accuracy: 0.8776
Epoch 14/300
29/31 [===========================>..] - ETA: 0s - loss: 0.1522 - accuracy: 0.8825
Yes You are here 0.8813387155532837 0.8897958993911743
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:02:47        17192
config.json            

 1/31 [..............................] - ETA: 0s - loss: 0.0966 - accuracy: 0.9062
Yes You are here 0.9208924770355225 0.9306122660636902
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:02:49        17192
config.json                                    2023-05-17 08:02:49         1588
metadata.json                                  2023-05-17 08:02:49           64
31/31 [==============================] - 0s 3ms/step - loss: 0.0876 - accuracy: 0.9209 - val_loss: 0.0996 - val_accuracy: 0.9306
Epoch 35/300
31/31 [================

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:02:52        17192
config.json                                    2023-05-17 08:02:52         1588
metadata.json                                  2023-05-17 08:02:52           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0577 - accuracy: 0.9533 - val_loss: 0.0672 - val_accuracy: 0.9592
Epoch 60/300
 1/31 [..............................] - ETA: 0s - loss: 0.0481 - accuracy: 1.0000
Yes You are here 0.9563894271850586 0.9632652997970581
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name              

29/31 [===========================>..] - ETA: 0s - loss: 0.0404 - accuracy: 0.9709
Yes You are here 0.9726166129112244 0.9755101799964905
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:02:55        17192
config.json                                    2023-05-17 08:02:55         1588
metadata.json                                  2023-05-17 08:02:55           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0397 - accuracy: 0.9726 - val_loss: 0.0479 - val_accuracy: 0.9755
Epoch 92/300
31/31 [================

 1/31 [..............................] - ETA: 0s - loss: 0.0324 - accuracy: 0.9688
Yes You are here 0.9736308455467224 0.9795918464660645
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:02:57        17192
config.json                                    2023-05-17 08:02:57         1588
metadata.json                                  2023-05-17 08:02:57           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0330 - accuracy: 0.9736 - val_loss: 0.0412 - val_accuracy: 0.9796
Epoch 114/300
31/31 [===============

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:02:59        17192
config.json                                    2023-05-17 08:02:59         1588
metadata.json                                  2023-05-17 08:02:59           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0283 - accuracy: 0.9746 - val_loss: 0.0362 - val_accuracy: 0.9837
Epoch 134/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0291 - accuracy: 0.9746 - val_loss: 0.0372 - val_accuracy: 0.9755
Epoch 135/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0280 - accuracy: 0.9757 - val_loss: 0.0361 - val_accuracy: 0.9796
Epoch 136/300
 1/31 [..............................] - ETA: 0s - loss: 0.0110 - accuracy: 1.0000
Yes You are here 0.9787018299102783 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
...........

metadata.json                                  2023-05-17 08:03:01           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0264 - accuracy: 0.9767 - val_loss: 0.0338 - val_accuracy: 0.9837
Epoch 149/300
29/31 [===========================>..] - ETA: 0s - loss: 0.0250 - accuracy: 0.9784
Yes You are here 0.9787018299102783 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:01        17192
config.json                                    2023-05-17 08:03:01         1588
metadata.json         

metadata.json                                  2023-05-17 08:03:02           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0244 - accuracy: 0.9746 - val_loss: 0.0336 - val_accuracy: 0.9837
Epoch 160/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0251 - accuracy: 0.9807 - val_loss: 0.0323 - val_accuracy: 0.9796
Epoch 161/300
31/31 [==============================] - ETA: 0s - loss: 0.0249 - accuracy: 0.9777
Yes You are here 0.9776875972747803 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                           

variables.h5                                   2023-05-17 08:03:04        17192
config.json                                    2023-05-17 08:03:04         1588
metadata.json                                  2023-05-17 08:03:04           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0215 - accuracy: 0.9817 - val_loss: 0.0311 - val_accuracy: 0.9837
Epoch 178/300
 1/31 [..............................] - ETA: 0s - loss: 0.0158 - accuracy: 0.9688
Yes You are here 0.9787018299102783 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5          

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:06        17192
config.json                                    2023-05-17 08:03:06         1588
metadata.json                                  2023-05-17 08:03:06           64
31/31 [==============================] - 0s 5ms/step - loss: 0.0209 - accuracy: 0.9757 - val_loss: 0.0306 - val_accuracy: 0.9837
Epoch 190/300
24/31 [======================>.......] - ETA: 0s - loss: 0.0201 - accuracy: 0.9844
Yes You are here 0.982758641242981 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name              

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:07        17192
config.json                                    2023-05-17 08:03:07         1588
metadata.json                                  2023-05-17 08:03:07           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0194 - accuracy: 0.9787 - val_loss: 0.0300 - val_accuracy: 0.9837
Epoch 201/300
 1/31 [..............................] - ETA: 0s - loss: 0.0167 - accuracy: 1.0000
Yes You are here 0.9807302355766296 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name             

...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:08        17192
config.json                                    2023-05-17 08:03:08         1588
metadata.json                                  2023-05-17 08:03:08           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0192 - accuracy: 0.9777 - val_loss: 0.0307 - val_accuracy: 0.9837
Epoch 212/300
31/31 [==============================] - 0s 2ms/step - loss: 0.0190 - accuracy: 0.9787 - val_loss: 0.0299 - val_accuracy: 0.9796
Epoch 213/300
 1/31 [..............................] - ETA: 0s - loss: 0.0244 - accuracy: 0.9688
Yes You are here 0.9787018299102783 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper


File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:10        17192
config.json                                    2023-05-17 08:03:10         1588
metadata.json                                  2023-05-17 08:03:10           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0186 - accuracy: 0.9828 - val_loss: 0.0298 - val_accuracy: 0.9837
Epoch 224/300
 1/31 [..............................] - ETA: 0s - loss: 0.0125 - accuracy: 1.0000
Yes You are here 0.9807302355766296 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name             

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:11        17192
config.json                                    2023-05-17 08:03:11         1588
metadata.json                                  2023-05-17 08:03:11           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0177 - accuracy: 0.9828 - val_loss: 0.0285 - val_accuracy: 0.9837
Epoch 235/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0173 - accuracy: 0.9817 - val_loss: 0.0309 - val_accuracy: 0.9796
Epoch 236/300
 1/31 [..............................] - ETA: 0s - loss: 0.0082 - accuracy: 1.0000
Yes You are here 0.9807302355766296 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
........

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:12        17192
config.json                                    2023-05-17 08:03:12         1588
metadata.json                                  2023-05-17 08:03:12           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0173 - accuracy: 0.9787 - val_loss: 0.0316 - val_accuracy: 0.9837
Epoch 246/300
28/31 [==========================>...] - ETA: 0s - loss: 0.0173 - accuracy: 0.9766
Yes You are here 0.9776875972747803 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name             

31/31 [==============================] - 0s 3ms/step - loss: 0.0165 - accuracy: 0.9817 - val_loss: 0.0294 - val_accuracy: 0.9796
Epoch 260/300
28/31 [==========================>...] - ETA: 0s - loss: 0.0158 - accuracy: 0.9833
Yes You are here 0.9837728142738342 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:14        17192
config.json                                    2023-05-17 08:03:14         1588
metadata.json                                  2023-05-17 08:03:14           64
31/31 [===============

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:16        17192
config.json                                    2023-05-17 08:03:16         1588
metadata.json                                  2023-05-17 08:03:16           64
31/31 [==============================] - 0s 5ms/step - loss: 0.0157 - accuracy: 0.9828 - val_loss: 0.0309 - val_accuracy: 0.9837
Epoch 274/300
 1/31 [..............................] - ETA: 0s - loss: 0.0156 - accuracy: 0.9688
Yes You are here 0.9837728142738342 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name             

File Name                                             Modified             Size
variables.h5                                   2023-05-17 08:03:17        17192
config.json                                    2023-05-17 08:03:17         1588
metadata.json                                  2023-05-17 08:03:17           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0148 - accuracy: 0.9848 - val_loss: 0.0312 - val_accuracy: 0.9837
Epoch 287/300
 1/31 [..............................] - ETA: 0s - loss: 0.0092 - accuracy: 1.0000
Yes You are here 0.9837728142738342 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name             

config.json                                    2023-05-17 08:03:18         1588
metadata.json                                  2023-05-17 08:03:18           64
31/31 [==============================] - 0s 4ms/step - loss: 0.0151 - accuracy: 0.9817 - val_loss: 0.0316 - val_accuracy: 0.9837
Epoch 297/300
31/31 [==============================] - 0s 3ms/step - loss: 0.0149 - accuracy: 0.9797 - val_loss: 0.0329 - val_accuracy: 0.9796
Epoch 298/300
 1/31 [..............................] - ETA: 0s - loss: 0.0075 - accuracy: 1.0000
Yes You are here 0.9858012199401855 0.9836734533309937
Keras weights file (<HDF5 file "variables.h5" (mode r+)>) saving:
...layers
......dense
.........vars
............0
............1
...metrics
......mean
.........vars
............0
............1
......mean_metric_wrapper
.........vars
............0
............1
...optimizer
......vars
.........0
.........1
.........2
.........3
.........4
...vars
Keras model archive saving:
File Name                              

In [13]:
accurate_train_accuracy,classify_report= test_model(model,X_train,Y_train)

31/31 [==============================] - 0s 813us/step


In [14]:
accurate_train_accuracy

0.9878296146044625

In [15]:
print(classify_report)

              precision    recall  f1-score   support

           0       0.95      1.00      0.97       127
           1       0.99      0.92      0.95       134
           2       0.99      1.00      0.99       271
           3       1.00      1.00      1.00       454

    accuracy                           0.99       986
   macro avg       0.98      0.98      0.98       986
weighted avg       0.99      0.99      0.99       986



In [16]:
accurate_val_accuracy,classify_report= test_model(model,X_val,Y_val)

8/8 [==============================] - 0s 1ms/step


In [17]:
accurate_val_accuracy

0.9836734693877551

In [18]:
print(classify_report)

              precision    recall  f1-score   support

           0       1.00      1.00      1.00        29
           1       1.00      0.97      0.98        31
           2       0.96      0.99      0.97        73
           3       0.99      0.98      0.99       112

    accuracy                           0.98       245
   macro avg       0.99      0.98      0.99       245
weighted avg       0.98      0.98      0.98       245



In [19]:
accurate_test_accuracy,classify_report= test_model(model,X_test,Y_test)

8/8 [==============================] - 0s 1ms/step


In [20]:
accurate_test_accuracy

0.9715447154471545

In [ ]:
print(classify_report)

## important-Note
gamma=1 on-test(0.967479674796748) on-val(0.9836734693877551)
gamma=2 on-test(0.967479674796748) on-val(0.9836734693877551) 
gamma=3 on-test(0.975609756097561) on-val(0.9836734693877551)

In [ ]:
## I can't consider observations in the  age to be false outlier 
## I will consider it a true outlier, I will just make a log transformation for it 
## bc it's so skewed , it have skewness value 1.481576 

# you can't do log transformation in the Time_E_DEV column 


# meal_Count log transforamtion has increase the skewness 





In [ ]:
df_original